File: cpp/145_BinaryTreePostorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ret;
        stack<TreeNode*> sTree;
        sTree.push(root);
        while(sTree.size()){
            auto t = sTree.top();
            if(t == NULL){
                sTree.pop();
                if(sTree.size()){
                   auto p = sTree.top();
                   sTree.pop();
                   if(p!= NULL) ret.push_back(p->val);
                }
            }else{
                if(t->left == NULL && t->right == NULL){
                    ret.push_back(t->val);
                    sTree.pop();
                }else{
                    sTree.push(NULL);
                    if(t->right) sTree.push(t->right);
                    if(t->left) sTree.push(t->left);
                }
            }
            
        }
        return ret;
    }
};if

File: js/145_BinaryTreePostorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function(root) {
    var ret = [];
    var stack = [];
    stack.push(root);
    while(stack.length > 0){
        var t = stack[stack.length-1];
        if(t === null){
            stack.pop();
            if(stack.length > 0){
                var p = stack.pop();
                if(p !== null) ret.push(p.val);
            }
        }else{
            if(t.left === null && t.right === null){
                ret.push(t.val);
                stack.pop();
            }else{
                stack.push(null);
                if(t.right !== null) stack.push(t.right);
                if(t.left !== null) stack.push(t.left);
            }
        }
    }
    return ret;
};